File: server/worker/src/database/projections/asset.rs
use chrono::{DateTime, Utc};
use evm::address::Address;
use fraction::Fraction;
use serde::{Deserialize, Serialize};
use sqlx::types::Uuid;

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub struct Asset {
    pub id: Uuid,
    pub created_at: DateTime<Utc>,
    pub last_modification_at: DateTime<Utc>,
    pub name: String,
    pub icon_path: String,
    pub symbol: String,
    pub address: Address,
    pub decimals: Fraction,
    pub maker_fee: Fraction,
    pub taker_fee: Fraction,
    pub transfer_fee: Fraction,
}

File: web/packages/types/asset.ts
import { z } from "zod";
import { Fraction } from "./primitives/fraction";
import { Datetime } from "./primitives/datetime";
import { Uuid } from "./primitives/uuid";
import { Address } from "@packages/components/providers/SessionProvider/models";

export const Asset = z.object({
  id: Uuid,
  created_at: Datetime,
  last_modification_at: Datetime,
  name: z.string(),
  icon_path: z.string(),
  symbol: z.string(),
  address: Address,
  decimals: Fraction,
  maker_fee: Fraction,
  taker_fee: Fraction,
  transfer_fee: Fraction,
});
export type Asset = z.infer<typeof Asset>;
